In [1]:
import pandas as pd

# Data Dictionary

# Data Dictionary – NYC Yellow Taxi Trip Records

| **Field Name** | **Description** |
|---------------|-----------------|
| **VendorID** | Code indicating the TPEP provider that supplied the trip record. <br>**1** = Creative Mobile Technologies (CMT) <br>**2** = Curb Mobility <br>**6** = Myle Technologies <br>**7** = Helix |
| **tpep_pickup_datetime** | Date and time when the taximeter was engaged (pickup time). |
| **tpep_dropoff_datetime** | Date and time when the taximeter was disengaged (dropoff time). |
| **passenger_count** | Number of passengers in the vehicle. |
| **trip_distance** | Trip distance reported by the taximeter, measured in **miles**. |
| **RatecodeID** | Final fare rate applied at the end of the trip. <br>**1** = Standard rate <br>**2** = JFK <br>**3** = Newark <br>**4** = Nassau or Westchester <br>**5** = Negotiated fare <br>**6** = Group ride <br>**99** = Null / Unknown |
| **store_and_fwd_flag** | Indicates whether the trip record was stored in the vehicle before being sent to the vendor due to lack of network connection. <br>**Y** = Store and forward <br>**N** = Sent directly |
| **PULocationID** | TLC Taxi Zone where the trip started. |
| **DOLocationID** | TLC Taxi Zone where the trip ended. |
| **payment_type** | Passenger payment type. <br>**0** = Flex Fare trip <br>**1** = Credit card <br>**2** = Cash <br>**3** = No charge <br>**4** = Dispute <br>**5** = Unknown <br>**6** = Voided trip |
| **fare_amount** | Time-and-distance fare calculated by the taximeter. |
| **extra** | Additional surcharges and miscellaneous charges. |
| **mta_tax** | Tax automatically applied based on the metered fare. |
| **tip_amount** | Tip amount recorded for **credit card payments only** (cash tips are not included). |
| **tolls_amount** | Total toll charges paid during the trip. |
| **improvement_surcharge** | Improvement surcharge applied at the start of the trip (introduced in 2015). |
| **total_amount** | Total amount charged to the passenger, excluding cash tips. |
| **congestion_surcharge** | NY State congestion surcharge collected during the trip. |
| **airport_fee** | Airport pickup fee, applied only for pickups at **LaGuardia (LGA)** and **John F. Kennedy (JFK)** airports. |
| **cbd_congestion_fee** | Per-trip charge for the MTA's **Congestion Relief Zone**, effective from **January 5, 2025**. |

---

## Additional Notes

- **Distance unit:** Miles
- **Time columns:** Stored as `datetime`.
- **`RatecodeID = 99`** represents **Null/Unknown** rather than a valid fare type.
- **`payment_type = 0`** represents **Flex Fare**, meaning the fare was determined through an upfront pricing mechanism rather than the standard metered fare.
- **`tip_amount`** only includes electronic (credit card) tips; cash tips are not recorded in the dataset.

# EDA DATA NĂM 2025

In [2]:
df1 = pd.read_parquet('../data/raw/yellow_tripdata_2025-01.parquet')
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3475226 entries, 0 to 3475225
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

In [3]:
# Kiểm tra các biến liên quan đến chi phí âm
monetary_cols = ['fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge', 'total_amount', 'congestion_surcharge', 'Airport_fee', 'cbd_congestion_fee']

In [4]:
negative_summary = pd.DataFrame({
    "negative_count": (df1[monetary_cols] < 0).sum(),
    "negative_percent": ((df1[monetary_cols] < 0).mean() * 100).round(2)
})

negative_summary.sort_values("negative_percent", ascending=False)

,negative_count,negative_percent
fare_amount,144118,4.15
total_amount,63037,1.81
improvement_surcharge,59530,1.71
mta_tax,57140,1.64
congestion_surcharge,48321,1.39
extra,29596,0.85
Airport_fee,10411,0.30
cbd_congestion_fee,6553,0.19
tolls_amount,4559,0.13
tip_amount,124,0.00


In [8]:
df1[df1["total_amount"] < 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
17,2,2025-01-01 00:01:41,2025-01-01 00:07:14,1.0,0.71,1.0,N,79,107,2,-7.20,-1.0,-0.5,3.66,0.0,-1.0,-8.54,-2.5,0.0,0.00
22,2,2025-01-01 00:55:54,2025-01-01 01:00:38,1.0,0.69,1.0,N,137,233,4,-6.50,-1.0,-0.5,0.00,0.0,-1.0,-11.50,-2.5,0.0,0.00
104,2,2025-01-01 00:56:12,2025-01-01 01:15:00,1.0,0.97,1.0,N,161,170,4,-16.30,-1.0,-0.5,0.00,0.0,-1.0,-21.30,-2.5,0.0,0.00
149,2,2025-01-01 00:55:53,2025-01-01 01:06:49,1.0,1.42,1.0,N,79,45,2,-12.10,-1.0,-0.5,0.00,0.0,-1.0,-17.10,-2.5,0.0,0.00
202,2,2025-01-01 00:29:35,2025-01-01 00:36:02,1.0,0.60,1.0,N,79,148,4,-7.20,-1.0,-0.5,0.00,0.0,-1.0,-12.20,-2.5,0.0,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3248636,2,2025-01-21 18:33:45,2025-01-21 18:51:10,NaN,2.25,NaN,None,231,107,0,-18.58,0.0,-0.5,0.00,0.0,-1.0,-23.33,NaN,NaN,0.00
3292466,2,2025-01-23 17:38:41,2025-01-23 17:46:47,NaN,0.00,NaN,None,137,186,0,-12.72,0.0,-0.5,0.00,0.0,-1.0,-17.47,NaN,NaN,0.00
3299623,2,2025-01-23 20:53:11,2025-01-23 21:04:29,NaN,3.90,NaN,None,233,232,0,-17.74,0.0,-0.5,0.00,0.0,-1.0,-22.49,NaN,NaN,0.00
3366645,2,2025-01-26 02:23:57,2025-01-26 02:47:25,NaN,4.82,NaN,None,255,188,0,-32.32,0.0,-0.5,0.00,0.0,-1.0,-33.82,NaN,NaN,0.00


In [9]:
df1['VendorID'].value_counts()

VendorID
2    2719860
1     753671
7       1206
6        489
Name: count, dtype: int64

In [10]:
# Check VendorID = 2
df1[(df1['VendorID'] == 2) & (df1['passenger_count'].isnull())]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
2935077,2,2025-01-01 00:39:17,2025-01-01 01:10:42,NaN,6.75,NaN,None,48,42,0,2.32,0.0,0.5,0.0,0.0,1.0,14.90,NaN,NaN,0.00
2935078,2,2025-01-01 00:33:40,2025-01-01 01:10:34,NaN,3.88,NaN,None,4,48,0,6.02,0.0,0.5,0.0,0.0,1.0,10.02,NaN,NaN,0.00
2935080,2,2025-01-01 00:21:42,2025-01-01 00:36:29,NaN,3.55,NaN,None,137,211,0,-2.35,0.0,0.5,0.0,0.0,1.0,5.80,NaN,NaN,0.00
2935081,2,2025-01-01 00:56:04,2025-01-01 01:06:11,NaN,1.55,NaN,None,140,233,0,5.06,0.0,0.5,0.0,0.0,1.0,9.06,NaN,NaN,0.00
2935083,2,2025-01-01 00:50:25,2025-01-01 00:57:34,NaN,1.04,NaN,None,142,239,0,14.16,0.0,0.5,2.0,0.0,1.0,20.16,NaN,NaN,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3475221,2,2025-01-31 23:01:48,2025-01-31 23:16:29,NaN,3.35,NaN,None,79,237,0,15.85,0.0,0.5,0.0,0.0,1.0,20.60,NaN,NaN,0.75
3475222,2,2025-01-31 23:50:29,2025-02-01 00:17:27,NaN,8.73,NaN,None,161,116,0,28.14,0.0,0.5,0.0,0.0,1.0,32.89,NaN,NaN,0.75
3475223,2,2025-01-31 23:26:59,2025-01-31 23:43:01,NaN,2.64,NaN,None,144,246,0,14.91,0.0,0.5,0.0,0.0,1.0,19.66,NaN,NaN,0.75
3475224,2,2025-01-31 23:14:34,2025-01-31 23:34:52,NaN,3.16,NaN,None,142,107,0,17.55,0.0,0.5,0.0,0.0,1.0,22.30,NaN,NaN,0.75


In [11]:
#Check vendor
# Giữ các dòng không có missing value ở bất kỳ cột nào
complete_rows = df1[df1.notna().all(axis=1)]

# Xem các VendorID xuất hiện trong những dòng hoàn chỉnh
complete_rows["VendorID"].unique()

complete_rows.groupby("VendorID").size()

VendorID
1     665467
2    2268404
7       1206
dtype: int64

=> Các dòng missing values thuộc vendorid = 2 và 6

In [12]:
df1[df1['VendorID'] == 6].isnull().sum()

VendorID                   0
tpep_pickup_datetime       0
tpep_dropoff_datetime      0
passenger_count          489
trip_distance              0
RatecodeID               489
store_and_fwd_flag       489
PULocationID               0
DOLocationID               0
payment_type               0
fare_amount                0
extra                      0
mta_tax                    0
tip_amount                 0
tolls_amount               0
improvement_surcharge      0
total_amount               0
congestion_surcharge     489
Airport_fee              489
cbd_congestion_fee         0
dtype: int64

=> Toàn bộ các dòng có vendorId = 6 đều chứa missing values

In [13]:
df1.isnull().sum()

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          540149
trip_distance                 0
RatecodeID               540149
store_and_fwd_flag       540149
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     540149
Airport_fee              540149
cbd_congestion_fee            0
dtype: int64

=> Nhận xét:
- Các file data năm 2025 đều có missing tại các biến giống nhau và số lượng missing của từng biến trong cùng 1 data bằng nhau
- Missing data tại các biến: passenger_count, RatecodeID, store_and_fwd_flag, congestion_surcharge, Airport_fee

In [14]:
df1.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,1,2025-01-01 00:18:38,2025-01-01 00:26:59,1.0,1.60,1.0,N,229,237,1,10.0,3.5,0.5,3.00,0.0,1.0,18.00,2.5,0.0,0.0
1,1,2025-01-01 00:32:40,2025-01-01 00:35:13,1.0,0.50,1.0,N,236,237,1,5.1,3.5,0.5,2.02,0.0,1.0,12.12,2.5,0.0,0.0
2,1,2025-01-01 00:44:04,2025-01-01 00:46:01,1.0,0.60,1.0,N,141,141,1,5.1,3.5,0.5,2.00,0.0,1.0,12.10,2.5,0.0,0.0
3,2,2025-01-01 00:14:27,2025-01-01 00:20:01,3.0,0.52,1.0,N,244,244,2,7.2,1.0,0.5,0.00,0.0,1.0,9.70,0.0,0.0,0.0
4,2,2025-01-01 00:21:34,2025-01-01 00:25:06,3.0,0.66,1.0,N,244,116,2,5.8,1.0,0.5,0.00,0.0,1.0,8.30,0.0,0.0,0.0


In [15]:
df1['payment_type'].value_counts()

payment_type
1    2444393
0     540149
2     390429
4      76481
3      23773
5          1
Name: count, dtype: int64

## Check missing values

In [16]:
# passenger_count
df1[df1['passenger_count']==0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
6,1,2025-01-01 00:14:47,2025-01-01 00:16:15,0.0,0.4,1.0,N,170,170,1,4.4,3.50,0.5,2.35,0.0,1.0,11.75,2.5,0.0,0.00
7,1,2025-01-01 00:39:27,2025-01-01 00:51:51,0.0,1.6,1.0,N,234,148,1,12.1,3.50,0.5,2.00,0.0,1.0,19.10,2.5,0.0,0.00
8,1,2025-01-01 00:53:43,2025-01-01 01:13:23,0.0,2.8,1.0,N,148,170,1,19.1,3.50,0.5,3.00,0.0,1.0,27.10,2.5,0.0,0.00
94,1,2025-01-01 00:11:27,2025-01-01 00:16:58,0.0,0.7,1.0,N,144,211,1,7.2,3.50,0.5,0.00,0.0,1.0,12.20,2.5,0.0,0.00
95,1,2025-01-01 00:19:30,2025-01-01 00:27:25,0.0,1.0,1.0,N,211,158,1,9.3,3.50,0.5,2.85,0.0,1.0,17.15,2.5,0.0,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2933984,1,2025-01-31 23:20:45,2025-01-31 23:45:28,0.0,3.9,1.0,N,144,50,1,25.4,4.25,0.5,2.50,0.0,1.0,33.65,2.5,0.0,0.75
2934049,1,2025-01-31 23:54:49,2025-02-01 00:09:35,0.0,6.4,1.0,N,263,232,1,28.2,4.25,0.5,2.00,0.0,1.0,35.95,2.5,0.0,0.75
2934589,1,2025-01-31 23:16:04,2025-01-31 23:36:38,0.0,2.4,1.0,N,68,114,1,15.6,4.25,0.5,4.25,0.0,1.0,25.60,2.5,0.0,0.75
2934590,1,2025-01-31 23:57:51,2025-02-01 00:08:06,0.0,1.6,1.0,N,142,236,1,10.7,3.50,0.5,3.10,0.0,1.0,18.80,2.5,0.0,0.00


In [19]:
df1[df1['passenger_count'].isnull()]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
2935077,2,2025-01-01 00:39:17,2025-01-01 01:10:42,NaN,6.75,NaN,None,48,42,0,2.32,0.0,0.5,0.0,0.0,1.0,14.90,NaN,NaN,0.00
2935078,2,2025-01-01 00:33:40,2025-01-01 01:10:34,NaN,3.88,NaN,None,4,48,0,6.02,0.0,0.5,0.0,0.0,1.0,10.02,NaN,NaN,0.00
2935079,1,2025-01-01 00:09:32,2025-01-01 00:12:38,NaN,0.00,NaN,None,158,158,0,22.72,0.0,0.5,0.0,0.0,1.0,26.72,NaN,NaN,0.00
2935080,2,2025-01-01 00:21:42,2025-01-01 00:36:29,NaN,3.55,NaN,None,137,211,0,-2.35,0.0,0.5,0.0,0.0,1.0,5.80,NaN,NaN,0.00
2935081,2,2025-01-01 00:56:04,2025-01-01 01:06:11,NaN,1.55,NaN,None,140,233,0,5.06,0.0,0.5,0.0,0.0,1.0,9.06,NaN,NaN,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3475221,2,2025-01-31 23:01:48,2025-01-31 23:16:29,NaN,3.35,NaN,None,79,237,0,15.85,0.0,0.5,0.0,0.0,1.0,20.60,NaN,NaN,0.75
3475222,2,2025-01-31 23:50:29,2025-02-01 00:17:27,NaN,8.73,NaN,None,161,116,0,28.14,0.0,0.5,0.0,0.0,1.0,32.89,NaN,NaN,0.75
3475223,2,2025-01-31 23:26:59,2025-01-31 23:43:01,NaN,2.64,NaN,None,144,246,0,14.91,0.0,0.5,0.0,0.0,1.0,19.66,NaN,NaN,0.75
3475224,2,2025-01-31 23:14:34,2025-01-31 23:34:52,NaN,3.16,NaN,None,142,107,0,17.55,0.0,0.5,0.0,0.0,1.0,22.30,NaN,NaN,0.75


=> Bên cạnh những dòng có 'passenger_count" là null, có 24 656 dòng có giá trị bằng 0

In [20]:
df1['RatecodeID'].value_counts(dropna=False)

RatecodeID
1.0     2756472
NaN      540149
2.0       94420
99.0      41963
5.0       26501
3.0        8622
4.0        7092
6.0           7
Name: count, dtype: int64

In [21]:
pd.crosstab(df1["VendorID"], df1["RatecodeID"].isna(), normalize="index")

RatecodeID,False,True
VendorID,,
1,0.882968,0.117032
2,0.834015,0.165985
6,0.000000,1.000000
7,1.000000,0.000000


In [22]:
df1[df1['RatecodeID'].isnull()]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
2935077,2,2025-01-01 00:39:17,2025-01-01 01:10:42,NaN,6.75,NaN,None,48,42,0,2.32,0.0,0.5,0.0,0.0,1.0,14.90,NaN,NaN,0.00
2935078,2,2025-01-01 00:33:40,2025-01-01 01:10:34,NaN,3.88,NaN,None,4,48,0,6.02,0.0,0.5,0.0,0.0,1.0,10.02,NaN,NaN,0.00
2935079,1,2025-01-01 00:09:32,2025-01-01 00:12:38,NaN,0.00,NaN,None,158,158,0,22.72,0.0,0.5,0.0,0.0,1.0,26.72,NaN,NaN,0.00
2935080,2,2025-01-01 00:21:42,2025-01-01 00:36:29,NaN,3.55,NaN,None,137,211,0,-2.35,0.0,0.5,0.0,0.0,1.0,5.80,NaN,NaN,0.00
2935081,2,2025-01-01 00:56:04,2025-01-01 01:06:11,NaN,1.55,NaN,None,140,233,0,5.06,0.0,0.5,0.0,0.0,1.0,9.06,NaN,NaN,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3475221,2,2025-01-31 23:01:48,2025-01-31 23:16:29,NaN,3.35,NaN,None,79,237,0,15.85,0.0,0.5,0.0,0.0,1.0,20.60,NaN,NaN,0.75
3475222,2,2025-01-31 23:50:29,2025-02-01 00:17:27,NaN,8.73,NaN,None,161,116,0,28.14,0.0,0.5,0.0,0.0,1.0,32.89,NaN,NaN,0.75
3475223,2,2025-01-31 23:26:59,2025-01-31 23:43:01,NaN,2.64,NaN,None,144,246,0,14.91,0.0,0.5,0.0,0.0,1.0,19.66,NaN,NaN,0.75
3475224,2,2025-01-31 23:14:34,2025-01-31 23:34:52,NaN,3.16,NaN,None,142,107,0,17.55,0.0,0.5,0.0,0.0,1.0,22.30,NaN,NaN,0.75


In [23]:
df1['store_and_fwd_flag'].value_counts(dropna = False)

store_and_fwd_flag
N       2927431
None     540149
Y          7646
Name: count, dtype: int64

In [24]:
df1['congestion_surcharge'].value_counts(dropna = False)

congestion_surcharge
 2.5    2660818
 NaN     540149
 0.0     225938
-2.5      48321
Name: count, dtype: int64

Nhận xét: 
- congestion_surcharge là khoản phụ thu phí ùn tắc giao thông được cộng vào tiền cước taxi cho mỗi chuyến đi ở NYC
- Trong thực tế, giá trị của khoản thu này thường là 0 và 2.5. Tuy nhiên cũng có trường hợp giá trị là -2.5 khi điều chỉnh kế toán hoặc hoàn tiền (chuyến hủy, tính sai phụ phí, xử lý lại). 
- Đi sâu hơn vào những dòng có giá trị âm và NAN

In [25]:
df1[df1['congestion_surcharge'] < 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
17,2,2025-01-01 00:01:41,2025-01-01 00:07:14,1.0,0.71,1.0,N,79,107,2,-7.2,-1.0,-0.5,3.66,0.0,-1.0,-8.54,-2.5,0.0,0.00
22,2,2025-01-01 00:55:54,2025-01-01 01:00:38,1.0,0.69,1.0,N,137,233,4,-6.5,-1.0,-0.5,0.00,0.0,-1.0,-11.50,-2.5,0.0,0.00
104,2,2025-01-01 00:56:12,2025-01-01 01:15:00,1.0,0.97,1.0,N,161,170,4,-16.3,-1.0,-0.5,0.00,0.0,-1.0,-21.30,-2.5,0.0,0.00
149,2,2025-01-01 00:55:53,2025-01-01 01:06:49,1.0,1.42,1.0,N,79,45,2,-12.1,-1.0,-0.5,0.00,0.0,-1.0,-17.10,-2.5,0.0,0.00
202,2,2025-01-01 00:29:35,2025-01-01 00:36:02,1.0,0.60,1.0,N,79,148,4,-7.2,-1.0,-0.5,0.00,0.0,-1.0,-12.20,-2.5,0.0,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2934953,2,2025-01-31 23:31:24,2025-01-31 23:34:04,1.0,0.51,1.0,N,162,170,2,-5.1,-1.0,-0.5,0.00,0.0,-1.0,-10.85,-2.5,0.0,-0.75
2934960,2,2025-01-31 23:41:22,2025-01-31 23:43:05,1.0,0.04,1.0,N,263,263,3,-3.7,-1.0,-0.5,0.00,0.0,-1.0,-8.70,-2.5,0.0,0.00
2934985,2,2025-01-31 23:03:01,2025-01-31 23:04:47,1.0,0.40,1.0,N,236,236,4,-4.4,-1.0,-0.5,0.00,0.0,-1.0,-9.40,-2.5,0.0,0.00
2935015,2,2025-01-31 23:14:54,2025-01-31 23:23:48,2.0,1.08,1.0,N,114,158,3,-9.3,-1.0,-0.5,0.00,0.0,-1.0,-15.05,-2.5,0.0,-0.75


In [26]:
df1[df1['congestion_surcharge'].isnull()]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
2935077,2,2025-01-01 00:39:17,2025-01-01 01:10:42,NaN,6.75,NaN,None,48,42,0,2.32,0.0,0.5,0.0,0.0,1.0,14.90,NaN,NaN,0.00
2935078,2,2025-01-01 00:33:40,2025-01-01 01:10:34,NaN,3.88,NaN,None,4,48,0,6.02,0.0,0.5,0.0,0.0,1.0,10.02,NaN,NaN,0.00
2935079,1,2025-01-01 00:09:32,2025-01-01 00:12:38,NaN,0.00,NaN,None,158,158,0,22.72,0.0,0.5,0.0,0.0,1.0,26.72,NaN,NaN,0.00
2935080,2,2025-01-01 00:21:42,2025-01-01 00:36:29,NaN,3.55,NaN,None,137,211,0,-2.35,0.0,0.5,0.0,0.0,1.0,5.80,NaN,NaN,0.00
2935081,2,2025-01-01 00:56:04,2025-01-01 01:06:11,NaN,1.55,NaN,None,140,233,0,5.06,0.0,0.5,0.0,0.0,1.0,9.06,NaN,NaN,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3475221,2,2025-01-31 23:01:48,2025-01-31 23:16:29,NaN,3.35,NaN,None,79,237,0,15.85,0.0,0.5,0.0,0.0,1.0,20.60,NaN,NaN,0.75
3475222,2,2025-01-31 23:50:29,2025-02-01 00:17:27,NaN,8.73,NaN,None,161,116,0,28.14,0.0,0.5,0.0,0.0,1.0,32.89,NaN,NaN,0.75
3475223,2,2025-01-31 23:26:59,2025-01-31 23:43:01,NaN,2.64,NaN,None,144,246,0,14.91,0.0,0.5,0.0,0.0,1.0,19.66,NaN,NaN,0.75
3475224,2,2025-01-31 23:14:34,2025-01-31 23:34:52,NaN,3.16,NaN,None,142,107,0,17.55,0.0,0.5,0.0,0.0,1.0,22.30,NaN,NaN,0.75


In [27]:
df1['Airport_fee'].value_counts(dropna= False)

Airport_fee
 0.00    2706446
 NaN      540149
 1.75     218203
-1.75      10411
 1.25          8
 5.00          7
 0.75          1
 6.75          1
Name: count, dtype: int64

## Lỗi Logic

In [29]:
# Lỗi giờ đón khách sau giờ trả khách 
df1[df1["tpep_pickup_datetime"] > df1["tpep_dropoff_datetime"]]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
99620,1,2025-01-02 12:26:00,2025-01-02 11:29:58,1.0,9.00,99.0,N,88,210,1,45.5,0.0,0.5,0.0,0.00,0.0,46.00,0.0,0.0,0.0
459885,1,2025-01-06 16:00:00,2025-01-06 15:05:30,1.0,3.80,99.0,N,208,254,1,24.5,0.0,0.5,0.0,0.00,0.0,25.00,0.0,0.0,0.0
1315880,1,2025-01-15 15:00:00,2025-01-15 14:42:48,1.0,1.00,99.0,N,107,4,1,18.5,0.0,0.5,0.0,0.00,0.0,19.00,0.0,0.0,0.0
2029534,1,2025-01-23 01:44:59,2024-12-18 07:52:40,0.0,3.10,1.0,N,162,238,1,21.9,2.5,0.5,4.5,0.00,1.0,30.40,2.5,0.0,0.0
2058555,1,2025-01-23 12:30:00,2025-01-23 11:44:59,1.0,3.90,99.0,N,236,116,1,27.5,0.0,0.5,0.0,0.00,0.0,28.00,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3419203,6,2025-01-29 22:01:23,2025-01-29 22:01:13,NaN,4.92,NaN,None,61,76,0,2.9,0.0,0.5,0.0,0.00,0.3,22.64,NaN,NaN,0.0
3424926,6,2025-01-30 08:01:40,2025-01-30 08:01:09,NaN,13.77,NaN,None,146,175,0,2.9,0.0,0.5,0.0,0.00,0.3,44.72,NaN,NaN,0.0
3430705,6,2025-01-30 16:01:38,2025-01-30 16:01:31,NaN,5.18,NaN,None,142,42,0,2.9,0.0,0.5,0.0,0.00,0.3,27.81,NaN,NaN,0.0
3443840,6,2025-01-30 23:01:38,2025-01-30 23:01:26,NaN,7.30,NaN,None,183,119,0,2.9,0.0,0.5,0.0,0.00,0.3,24.20,NaN,NaN,0.0


In [30]:
df1.duplicated().sum()

np.int64(0)

In [31]:
df1['payment_type'].value_counts()

payment_type
1    2444393
0     540149
2     390429
4      76481
3      23773
5          1
Name: count, dtype: int64

In [32]:
# Check giá trị âm fare_amount
df1[df1['fare_amount'] < 0]

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
17,2,2025-01-01 00:01:41,2025-01-01 00:07:14,1.0,0.71,1.0,N,79,107,2,-7.20,-1.0,-0.5,3.66,0.0,-1.0,-8.54,-2.5,0.0,0.00
22,2,2025-01-01 00:55:54,2025-01-01 01:00:38,1.0,0.69,1.0,N,137,233,4,-6.50,-1.0,-0.5,0.00,0.0,-1.0,-11.50,-2.5,0.0,0.00
104,2,2025-01-01 00:56:12,2025-01-01 01:15:00,1.0,0.97,1.0,N,161,170,4,-16.30,-1.0,-0.5,0.00,0.0,-1.0,-21.30,-2.5,0.0,0.00
149,2,2025-01-01 00:55:53,2025-01-01 01:06:49,1.0,1.42,1.0,N,79,45,2,-12.10,-1.0,-0.5,0.00,0.0,-1.0,-17.10,-2.5,0.0,0.00
202,2,2025-01-01 00:29:35,2025-01-01 00:36:02,1.0,0.60,1.0,N,79,148,4,-7.20,-1.0,-0.5,0.00,0.0,-1.0,-12.20,-2.5,0.0,0.00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3475198,2,2025-01-31 23:30:26,2025-01-31 23:36:41,NaN,1.21,NaN,None,49,17,0,-1.50,0.0,0.5,0.00,0.0,1.0,1.00,NaN,NaN,0.00
3475204,2,2025-01-31 23:33:05,2025-01-31 23:46:21,NaN,2.55,NaN,None,144,170,0,-4.75,0.0,0.5,0.00,0.0,1.0,4.91,NaN,NaN,0.75
3475205,2,2025-01-31 23:43:55,2025-01-31 23:48:40,NaN,0.45,NaN,None,231,45,0,-4.75,0.0,0.5,0.00,0.0,1.0,3.00,NaN,NaN,0.75
3475212,2,2025-01-31 23:00:46,2025-01-31 23:12:08,NaN,1.51,NaN,None,230,229,0,-4.75,0.0,0.5,0.00,0.0,1.0,2.07,NaN,NaN,0.75


Tuy nhiên khi giá trị fare_amount âm, các giá trị của biến khác liên quan đến phí cũng âm

In [33]:
df1.sort_values(by='fare_amount', ascending=False).head(10)

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
1780915,1,2025-01-20 12:07:18,2025-01-20 12:12:42,1.0,1.60,1.0,N,138,8,4,863372.12,6.75,0.5,0.0,0.00,1.0,863380.37,0.0,1.75,0.00
1404958,2,2025-01-16 12:23:14,2025-01-17 14:41:56,1.0,255.33,4.0,N,161,265,2,2450.90,0.00,0.0,0.0,54.06,1.0,2506.71,0.0,0.00,0.75
1832645,2,2025-01-21 05:17:57,2025-01-21 08:19:45,1.0,188.88,4.0,N,247,265,2,1309.20,1.00,0.5,0.0,0.00,1.0,1311.70,0.0,0.00,0.00
2358832,2,2025-01-26 00:04:04,2025-01-26 00:04:42,4.0,0.00,5.0,N,142,142,2,950.00,0.00,0.0,0.0,0.00,1.0,953.50,2.5,0.00,0.00
2358836,2,2025-01-26 00:22:47,2025-01-26 00:22:55,4.0,0.00,5.0,N,226,226,2,950.00,0.00,0.0,0.0,0.00,1.0,951.00,0.0,0.00,0.00
2358837,2,2025-01-26 00:24:25,2025-01-26 00:24:35,4.0,0.00,5.0,N,226,226,2,950.00,0.00,0.0,0.0,0.00,1.0,951.00,0.0,0.00,0.00
1346041,2,2025-01-15 19:14:42,2025-01-15 21:46:13,1.0,143.54,4.0,N,132,265,1,936.80,2.50,0.0,0.0,27.00,1.0,969.05,0.0,1.75,0.00
562143,2,2025-01-07 19:12:25,2025-01-07 19:14:04,1.0,0.10,5.0,N,226,226,3,900.00,0.00,0.0,0.0,0.00,1.0,901.00,0.0,0.00,0.00
2358828,2,2025-01-26 00:00:56,2025-01-26 00:01:06,4.0,0.00,5.0,N,142,142,2,900.00,0.00,0.0,0.0,0.00,1.0,903.50,2.5,0.00,0.00
2353880,2,2025-01-25 23:56:58,2025-01-25 23:57:06,4.0,0.00,5.0,N,142,142,2,899.99,0.00,0.0,0.0,0.00,1.0,903.49,2.5,0.00,0.00


Có những chuyến taxi có khoảng cách đi rất nhỏ nhưng fare_amount lại rất lớn => Cần xử lí outlier

In [34]:
df1['payment_type'].value_counts()

payment_type
1    2444393
0     540149
2     390429
4      76481
3      23773
5          1
Name: count, dtype: int64

In [37]:
# 1. Khoảng cách gần như 0 nhưng fare rất cao
mask_high_fare_short = (
    (df1["trip_distance"] < 0.1) &
    (df1["fare_amount"] > 300)
)

# 2. Khoảng cách = 0, fare = 0 nhưng khác zone
mask_zero_trip_diff_zone = (
    (df1["trip_distance"] == 0) &
    (df1["fare_amount"] == 0) &
    (df1["PULocationID"] != df1["DOLocationID"])
)

# 3. Chuyến đi cực dài
mask_extreme_distance = df1["trip_distance"] > 250

# 4. Hành khách quá nhiều
mask_passenger = df1["passenger_count"] > 6

#5. Thời gian đón khách sau thời gian trả khách 
mask_false_pickup_logic = df1['tpep_pickup_datetime'] > df1['tpep_dropoff_datetime']

In [38]:
mask_false_pickup_logic.sum()

np.int64(124)

In [39]:
outlier_summary = pd.DataFrame({
    "Outlier": [
        "High fare, short distance",
        "Zero trip, different zone",
        "Extreme distance",
        "Passenger > 6",
        "PickUp > DropOff"
    ],
    "Count": [
        mask_high_fare_short.sum(),
        mask_zero_trip_diff_zone.sum(),
        mask_extreme_distance.sum(),
        mask_passenger.sum(),
        mask_false_pickup_logic.sum()
    ],
    "Action": [
        "Remove",
        "Keep (record)",
        "Remove",
        "Remove",
        "Remove"
    ]
})

outlier_summary

,Outlier,Count,Action
0,"High fare, short distance",79,Remove
1,"Zero trip, different zone",133,Keep (record)
2,Extreme distance,120,Remove
3,Passenger > 6,18,Remove
4,PickUp > DropOff,124,Remove


In [40]:
# Đếm được 

In [41]:
# Gộp tất cả điều kiện cần loại bỏ
remove_mask = (
    mask_high_fare_short |
    mask_zero_trip_diff_zone |
    mask_extreme_distance |
    mask_passenger |
    mask_false_pickup_logic
)

# Loại bỏ outlier
df1_clean = df1.loc[~remove_mask].copy()

print(f"Rows before: {len(df1):,}")
print(f"Rows removed: {remove_mask.sum():,}")
print(f"Rows after: {len(df1_clean):,}")

Rows before: 3,475,226
Rows removed: 474
Rows after: 3,474,752


In [43]:
import seaborn as sns
import matplotlib.pyplot as plt

plt.figure(figsize=(12,10))
sns.heatmap(
    corr,
    annot=True,
    fmt=".2f",
    cmap="RdBu_r",
    center=0,
    square=True,
    linewidths=0.5
)
plt.title("Correlation Matrix")
plt.show()

NameError: name 'corr' is not defined

<Figure size 1200x1000 with 0 Axes>

- Loại 'total_amount' vì có correlation cao với fare_amount. Total_amount là tổng cộng các biến chi phí

### Nhận xét tổng:
1. Missing Values:
- 5 biến 'passenger_count', 'RatecodeID', 'store_and_fwd_flag', 'congestion_surcharge', 'Airport_fee' đồng thời có số lượng missing values bằng nhau và cùng 1 dòng.
- Biến 'RatecodeID' có giá trị null và có cả giá trị 99. Cần check lại giá trị 99 này vì theo một số nguồn nó được coi là ko xác định 
- Trong đó, các missing_values phụ thuộc vào VendorID, với 489 dòng có vendorID = 6 và đoạn cuối data (451456 dòng) với VendorId = 2
- congestion_surcharge và Airport_fee bị missing có thể coi là không có (tự cân nhắc)

2. Outlier/ Lỗi logic

Các quy tắc dưới đây được sử dụng để phát hiện các chuyến taxi có dữ liệu bất thường hoặc không hợp lý trước khi thực hiện làm sạch dữ liệu.

| **Loại bất thường** | **Điều kiện** |
|----------------------|---------------|
| Giá cước rất cao nhưng quãng đường gần như bằng 0 | `trip_distance < 0.1` và `fare_amount > 300` |
| Chuyến đi có quãng đường bằng 0, giá cước bằng 0 nhưng khác vùng đón và trả khách | `trip_distance = 0` và `fare_amount = 0` và `PULocationID ≠ DOLocationID` | 
| Chuyến đi có quãng đường bất thường | `trip_distance > 250` (mile) |
| Số hành khách vượt quá sức chứa của taxi | `passenger_count > 6` |
| Thời gian đón khách xảy ra sau thời gian trả khách | `tpep_pickup_datetime > tpep_dropoff_datetime` | 